File: apps/mobile/compose/src/commonMain/kotlin/co/typie/editor/Publication.kt
package co.typie.editor

import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.unit.IntSize
import co.typie.editor.ffi.FrameKey
import kotlin.jvm.JvmInline

@JvmInline internal value class SurfaceKey(val value: Long)

internal data class SurfaceTarget(
  val page: Int,
  val key: SurfaceKey,
  val configuration: SurfaceConfiguration,
)

internal data class FrameProof(
  val editorRevision: Long,
  val surfaceKey: SurfaceKey,
  val frameKey: FrameKey,
)

internal data class PresentedFrame(
  val bitmap: ImageBitmap,
  val pixelSize: IntSize,
  val proof: FrameProof,
)

internal object Publication {
  fun satisfiesWaiter(
    requestedRevision: Long,
    publishedRevision: Long?,
    frames: Map<Int, PresentedFrame>,
    requireFrame: Boolean = false,
  ): Boolean =
    publishedRevision != null &&
      publishedRevision >= requestedRevision &&
      (!requireFrame || frames.isNotEmpty())

  fun accepts(
    proof: FrameProof,
    target: SurfaceTarget,
    requiredRevision: Long?,
    available: Boolean,
  ): Boolean =
    available &&
      proof.surfaceKey == target.key &&
      (requiredRevision == null || proof.editorRevision >= requiredRevision)

  fun workRevision(appliedRevision: Long, requiredRevision: Long?): Long? = requiredRevision?.let {
    maxOf(appliedRevision, it)
  }
}

File: apps/website/src/lib/editor-ffi/publication.ts
export type FrameProof = {
  revision: number;
  surfaceKey: number;
  frameKey: number;
};

export type PublicationTarget = {
  key: number;
  requiredRevision: number | undefined;
  proof: FrameProof | undefined;
  available: boolean;
};

type PublishedFrameFacts = {
  surfaceKey: number;
};

export function proofSatisfies(target: PublicationTarget): boolean {
  if (!target.available || target.proof === undefined || target.proof.surfaceKey !== target.key) return false;
  return target.requiredRevision === undefined || target.proof.revision >= target.requiredRevision;
}

export function satisfiesWaiter(
  requestedRevision: number,
  publishedRevision: number | undefined,
  frames: ReadonlyMap<number, PublishedFrameFacts>,
  requireFrame = false,
): boolean {
  if (publishedRevision === undefined || publishedRevision < requestedRevision) return false;
  return !requireFrame || frames.size > 0;
}
